state/settings.go

```diff
@@ -266,6 +266,29 @@ func (s *Settings) applyChanges(changes settings.ItemChanges) {
 	}
 }
 
+// WriteBulkSettings writes all of the input settings documents
+// in a single transaction.
+// NOTE: The settings are passed as values and not references.
+// Unlike the call to "Write" on a single settings reference,
+// we cannot expect all the settings objects to be updated to represent
+// consistency between the live and on-disk values.
+// TODO (manadart 2019-04-30): This is inelegant.
+// The settings abstractions are poor, but model generations really makes
+// this logic creak.
+func (st *State) WriteBulkSettings(s []Settings) error {
```

Comment:
Given that the API surface for Settings is not that large couldn't you extract an interface out of it and define a type (e.g SettingsGroup) whose Write() implementation does the ops aggregation?

Reply:
That is how I approached it initially (minus the interface). It's in the first commit (https://github.com/juju/juju/pull/10120/commits/7ed76f88e42fab74cb774b1c5ce948171a1a1730).

But this type is not passed around, at least the with the current way I have done it. The only way an indirection would be of use is if we pushed the generation of the group into _state/modelgeneration.go_. I'd rather do that kind of logic at the API server level rather than in the `state` package, which already has too much business logic and interdependencies.

The blocker I ran into is that I couldn't test it (creating a mock `state.Database` and using it in the `Settings` instances) without exporting more of settings from the state package.

If you have an insight here, I'd be happy to hear it.

Reply:
Ah, I didn't see that because I was reviewing in the Files changed view.

Disclaimer: I don't know enough about settings so the following blurb may not apply...

Would it be possible to extract Settings out of the state package? Getting a DB handle from outside the state package is definitely a problem but if they could live in a standalone package perhaps you could have a `State.WriteSettings(Settings)` (where Settings is an interface)

Apart from this bit the code LGTM. Perhaps someone with more experience around this part of the codebase could suggest a more efficient approach to implementing this.

Reply:
I think what we really need here is something like `settingsUpdateOps` used as part of the build txn for `branch.Commit`. This way the read of the settings is part of the build txn and in sync, and the committing of the branch and the updating of the settings are all done in one database transaction.

Reply:
All done in the one transaction now, including assertions documents have not changed between reading and writing.